amm/contracts/router/lib.rs

```diff
@@ -47,9 +53,13 @@ pub mod router {
             token_0: AccountId,
             token_1: AccountId,
         ) -> Result<AccountId, RouterError> {
-            self.factory_ref()
-                .get_pair(token_0, token_1)
-                .ok_or(RouterError::PairNotFound)
+            if let Some(pair) = self.pairs.get((token_0, token_1)) {
+                Ok(pair)
+            } else {
+                self.factory_ref()
+                    .get_pair(token_0, token_1)
+                    .ok_or(RouterError::PairNotFound)
```

Comment:
This is a little strange. Is this branch of the `if` ever happening? If it is, we should probably add the result to cache.

Reply:
Yes, we should add the result to the cache.

Reply:
That was my first idea, but that requires making `get_pair` a storage-modifying message (`&mut self`) instead of read only (`&self`). That, in turns, forces quite a few other read-only methods to become storage-modifying. I thought it's safer to minimize the amount of such changes.

Reply:
Is this related to creating pair w/o going through the router? 

I agree we could cache but that will make a difference only for pairs created directly on the `Factory` itself.

Reply:
Ok, I guess it's fine. In the worst-case scenario, if for some reason the pairs are not cached for some reason, we can redeploy the router...